modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -459,17 +461,42 @@ private long absFileIdx(CheckpointEntry pointer) {
      * @param partsCounter Partition mapped to update counter.
      * @return Earliest WAL pointer for group specified.
      */
-    @Nullable public WALPointer searchEarliestWalPointer(int grpId, Map<Integer, Long> partsCounter) throws IgniteCheckedException {
+    @Nullable public WALPointer searchEarliestWalPointer(
+        int grpId,
+        Map<Integer, Long> partsCounter,
+        long margin
+    ) throws IgniteCheckedException {
         if (F.isEmpty(partsCounter))
             return null;
 
         Map<Integer, Long> modifiedPartsCounter = new HashMap<>(partsCounter);
 
         FileWALPointer minPtr = null;
 
+        LinkedList<T3<Map.Entry<Integer, Long>, FileWALPointer, Long>> historyPointerCandidat = new LinkedList<>();
+
         for (Long cpTs : checkpoints(true)) {
             CheckpointEntry cpEntry = entry(cpTs);
 
+            while (!F.isEmpty(historyPointerCandidat)) {
```

Comment:
This is **really** hard to read.
Can we extract this logic to separate method and add more comments on what logic is actually performed here?